hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/deltacommit/SparkUpsertDeltaCommitPartitioner.java

```diff
@@ -89,10 +89,13 @@ protected List<SmallFile> getSmallFiles(String partitionPath) {
   private List<FileSlice> getSmallFileCandidates(String partitionPath, HoodieInstant latestCommitInstant) {
     // If we can index log files, we can add more inserts to log files for fileIds NOT including those under
     // pending compaction
+    Comparator<FileSlice> comparator = Comparator.<FileSlice, Long>comparing(fileSlice -> getTotalFileSize(fileSlice))
+        .thenComparing(FileSlice::getFileId);
     if (table.getIndex().canIndexLogFiles()) {
       return table.getSliceView()
               .getLatestFileSlicesBeforeOrOn(partitionPath, latestCommitInstant.getTimestamp(), false)
               .filter(this::isSmallFile)
+              .sorted(comparator)
               .collect(Collectors.toList());
```

Comment:
Should we just fix the tests? Do we have gains for the sort in production?

Reply:
I think it makes sense to prefer the smallest files first as candidates to minimize IO

Reply:
Fine, just fix all the test failures and let's see what use cases are affected.

Reply:
@danny0405 the test failing is flakey. It is try to test some spark exception but it is non-deterministic.

Reply:
Looks like the test failure takes some time to fix, is it easy to make the tests deterministic?

Reply:
I've been trying but do not understand why these tests are coupled to small file handling. I think that the approach to testing is a bit strange to rely on that feature to test something in delta streamer for example.